### PR TITLE
Fix validation set configuration via CLI

### PR DESCRIPTION
**Pull request by ejnnr**
**The problem:** `train_classifier.py` accepts a dictionary of validation datasets as one of its config options. But `simple_parsing` doesn't support parsing dictionaries of dataclasses from the CLI (there's an [issue for lists](https://github.com/lebrice/SimpleParsing/issues/271), I think dictionaries would be trickier if anything because it's less clear what the syntax should be).

**Solution in this PR:** I introduced a new config dataclass `ValidationConfig` which serves as a replacement for the dictionary. The big downside is that it has a finite hard-coded list of possible names (i.e. dict keys) for validation splits (right now they're "val", "clean", and "backdoor"). By default they're all set to a new dummy config, and users can override any of them individually and change their properties etc. from the CLI.

This allows pretty complex validation sets, but the CLI usage quickly becomes verbose. For example, say we want to train on poisoned data, and evaluate performance on clean and backdoor data separately. That could look as follows:
```bash
python -m cupbearer.scripts.train_classifier --train_data backdoor --train_data.original mnist --train_data.backdoor corner --train_data.backdoor.p_backdoor 0.1 --model mlp --num_epochs 2 --val_data.backdoor backdoor --val_data.backdoor.original mnist --val_data.backdoor.backdoor corner --val_data.backdoor.original.train false --val_data.clean mnist --val_data.clean.train false
```
If we go with this approach, one improvement could be to add more special cases for `ValidationConfig` that copy reasonable defaults from the training data. The main change this would require is passing the training data config to `ValidationConfig` somehow. I could look into that but first we should figure out if this is even a reasonable approach overall.
Another thing that could help is just having more aliases, e.g. a separate "mnist_test" config option.

**Alternatives I considered:**
- Add rudimentary support for dicts of dataclasses to `simple_parsing`. This would perhaps be cleanest, but my sense is that it's non-trivial, and we'd then still have the problem of verbose CLI usage
- Just don't support a dict of validation datasets. We could have *only* some list of cases people usually want instead of allowing arbitrary custom configs here. Maybe this is the way to go, especially since the current approach also just supports a hard-coded list of split names. (It's trivial to add new ones though)
- Do our own hacky parsing, where the dict is specified as a string or list of (name, config_name) pairs, e.g. ("validation", "mnist"). This doesn't let you override any arguments to the DatasetConfigs though, e.g. the backdoor example above couldn't be specified.

### CODE REVIEW

**Comment by ejnnr** [comment]
@VRehnberg If you have ideas for fixing this more elegantly, I'd be very happy to use them instead of this. I don't think getting this exactly right is a priority at the moment, it should just be reasonably painless to work with for any experiments you want to do.

File: src/cupbearer/scripts/conf/train_classifier_conf.py [hunk 1]
```diff
@@ -1,21 +1,50 @@
...
-from simple_parsing.helpers import dict_field, mutable_field
+from cupbearer.utils.utils import BaseConfig
+from simple_parsing.helpers import mutable_field
+
+
+@dataclass(kw_only=True)
+class ValidationConfig(BaseConfig):
+    val: Optional[DatasetConfig] = config_group(DatasetConfig, NoData)
+    clean: Optional[DatasetConfig] = config_group(DatasetConfig, NoData)
+    backdoor: Optional[DatasetConfig] = config_group(DatasetConfig, NoData)
```
Comment (VRehnberg) [inline src/cupbearer/scripts/conf/train_classifier_conf.py:22]: As far as I can tell the names have no actual difference in what they actually imply besides convention.

On the other hand there are a lot of things you could wish for, but I think the best thing right now is just to have something good enough. As such having these names might be better than `val1`, `val2`, `val3`, ... as then the logs are more readable as long as you're following convention.

The following are the validation I've been wanting so far:
 - Regular validation dataset (same setting as training)
 - Clean, no backdoor, no adversarial attack, etc.
 - Backdoor, p_backdoor = 1.0 to see accuracy for backdoor detection
 - Alt. backdoor to use something like the backdoor but where backdoor label should not be used (only Wanet has this implemented right now, but I can see a use for at least corner as well)
---
For now I say this is good enough and we add any changes as needed.

**Review by VRehnberg — APPROVED** [review]
Lgtm
